### DISCUSSION TITLE
Application Insights for Dotnet MAUI

### DISCUSSION THREAD

**Discussion by wjvii** [dotnet/maui#317] [Ideas]
## Summary
We develop desktop, mobile and web applications using primarily WPF, Xamarin Forms, ASP.NET Core, Azure Functions, Blazor. We handle all our devops with Azure DevOps and due to the nature of our products also use SQLCipher with SQLite for our offline data. We love Application Insights and use it for every tier of our application where it is available, which means everything except Xamarin Forms. Since Application Insights is already cross platform, supports offline scenarios and more it would be great to have it directly supported by .NET MAUI on all platforms .NET MAUI will support. This would give developers a single telemetry platform to use across most if not all of the Microsoft .NET ecosystem and allow shops like ours to not have to integrate with AppCenter, not have to setup and use the AppCenter portal, and not have to use an entirely different SDK and API for telemetry. I realize that AppCenter can push data to Application Insights, but it would be great to just go direct and save the extra steps. Not to mention that AppCenter and applications with SQLCipher can be a bit problematic to get working correctly, at least that has been our experience.

## API Changes

Not sure, from the developer's point of view I say none required but the api would need to have offline support that some of the existing TelemetryChannels currently have.

## Intended Use Case
Would allow developers to use Application Insights in all of .NET MAUI applications just like they can do in all other .NET applications.

**Comment by freever** [comment]
@wjvii I have been using the Microsoft.ApplicationInsights nuget package in Xamarin apps for ages, it works fine.

**Comment by wjvii** [comment]
@freever, which package(s) and TelemetryChannel configuration did you use to support preserving telemetry in offline scenarios with Xamarin Forms? I tried some experiments with Microsoft.ApplicationInsights.WindowsServer.TelemetryChannel and never was able to get it to preserve the data when offline and send it when back online. It did work fine and send telemetry when online but I lost all telemetry when offline. I a WPF application it worked fine online and offline both.

Thank you.

**Comment by freever** [comment]
That's a good point, I had never worried too much about offline scenarios.

A true Xamarin / Maui AppInsights implementation would indeed be very
welcome. I agree with all your points about App Center.

**Comment by charlesroddie** [comment]
It does make sense to complete at least one of ApplicationInsights and App Center, and preferably both. Since App Center doesn't support all platforms, we need to use App Center for iOS/UWP/Android and ApplicationInsights for Mac. App Center also hasn't shown any signs of life for a year so it's not clear if its promise of a single place to manage app data will ever be finished.

**Comment by DamianSuess** [comment]
Any word on MAUI AppInsights? Even if it's a flexible one that's not tied to a subscription. If not, the old fashion custom service route suits me just fine

**Comment by VladislavAntonyuk** [comment]
https://vladislavantonyuk.github.io/articles/Adding-Application-Insights-to-.NET-MAUI-Application

**Reply by luke-barnett** [reply]
@VladislavAntonyuk what happens here when the application doesn't have consistent internet connectivity? I'm looking for a solution that will take into account a mobile application being used in remote areas.

**Reply by VladislavAntonyuk** [reply]
For offline work, you can use MetroLog. It allows you to store logs on a device.
as for ApplictionInsights, I'll need to check

**Reply by VladislavAntonyuk** [reply]
@luke-barnett this should give you a start point for offline work with application insights: https://stackoverflow.com/questions/53721396/can-application-insights-be-used-off-line

Idea is to store logs somewhere offline (on storage). When connection restored again you can upload logs

**Reply by jeremy-bridges** [reply]
Does App Insights still reject events older than 48 hours old? That would be a deal breaker here, regardless of how the app stores logs when offline.

**Comment by viswanath-pamarthi** [comment]
Also, the offline scenario applies in the instances where the app crashes. I believe Appcenter saves the crash log and uploads it to appcenter on next app launch.

**Comment by Tviljan** [comment]
You should check this out https://github.com/dhindrik/TinyInsights.Maui. Seems to work offline also by storing things to a file and sending when network available
